### PR TITLE
Fix various issues with vsphere networking

### PR DESCRIPTION
**Pull request by wupeka**
## Description of change
This fixes various issues with VSphere networking. Instead of internal/external devices in vsphere that were treated differently we now have a single code for adding a network device. cloud-init ENI/Netplan generation code was refactored to be used by vsphere. Start of work on instance-based firewalling.
There is a known bug with vsphere warning about duplicated MAC addresses, it will be addressed in a separate PR.

## QA steps
Deploy a vsphere instance with external-network set, verify that it works.

### CODE REVIEW

File: cloudconfig/cloudinit/cloudinit_centos.go [hunk 1]
```diff
@@ -254,3 +256,9 @@ func (cfg *centOSCloudConfig) addRequiredPackages() {
 //addPackageProxyCmds. Research if calling the same is fine.
 func (cfg *centOSCloudConfig) updateProxySettings(proxySettings proxy.Settings) {
 }
+
+// AddNetworkConfig is defined on the NetworkingConfig interface.
+// TODO(wpk) This has to be implemented for CentOS on VSphere to work properly!
```
Comment (nskaggs) [inline cloudconfig/cloudinit/cloudinit_centos.go]: Is this really something we can do / control?

File: cloudconfig/cloudinit/cloudinit_win.go [hunk 2]
```diff
@@ -120,3 +122,8 @@ func (cfg *windowsCloudConfig) addRequiredPackages() {
 // updateProxySettings is defined on the AdvancedPackagingConfig interface.
 func (cfg *windowsCloudConfig) updateProxySettings(proxy.Settings) {
 }
+
+// AddNetworkConfig is defined on the NetworkingConfig interface.
+func (cfg *windowsCloudConfig) AddNetworkConfig(interfaces []network.InterfaceInfo) error {
+	return nil
```
Comment (nskaggs) [inline cloudconfig/cloudinit/cloudinit_win.go]: Does this and centos return nice errors in these cases? Let's not leave the user hanging.

File: provider/vsphere/internal/vsphereclient/createvm_test.go [hunk 3]
```diff
@@ -384,6 +385,35 @@ func (s *clientSuite) TestCreateVirtualMachineRootDiskSize(c *gc.C) {
...
+	}
+	for i, test := range testData {
+		c.Logf("test #%d: MAC=%s expected %s", i, test.Mac, test.Result)
+		c.Check(VerifyMAC(test.Mac), gc.Equals, test.Result)
+	}
+}
+
+// GenerateMAC is random so we verify a few times if the MACs returned are
+// are correct.
+func (s *clientSuite) TestGenerateMAC(c *gc.C) {
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm_test.go]: These seems a little odd to run. We should feel confident generatemac should always generate a valid mac address. Our larger concern should be around verifymac, and we could explore edge cases to confirm.

I don't think we need this.

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 4]
```diff
@@ -480,6 +484,40 @@ func (c *Client) addNetworkDevice(
 	return &networkDevice, nil
 }
 
+// GenerateMAC generates a random hardware address that meets VMWare
+// requirements for MAC address: 00:50:56:XX:YY:ZZ where XX is between 00 and 3f.
+// https://pubs.vmware.com/vsphere-4-esx-vcenter/index.jsp?topic=/com.vmware.vsphere.server_configclassic.doc_41/esx_server_config/advanced_networking/c_setting_up_mac_addresses.html
+func GenerateMAC() (string, error) {
+	c, err := rand.Int(rand.Reader, big.NewInt(4194303))
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm.go]: Useful to note 4194303 == 3F:FF:FF == max value here explicitly.

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 5]
```diff
@@ -480,6 +484,40 @@ func (c *Client) addNetworkDevice(
...
+	if err != nil {
+		return "", err
+	}
+	r := c.Uint64()
+	return fmt.Sprintf("00:50:56:%.2x:%.2x:%.2x", (r>>16)&0xff, (r>>8)&0xff, r&0xff), nil
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm.go]: 00:50:56 is really a const specific to vmware. Can we define and use it instead everywhere?

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 6]
```diff
@@ -480,6 +484,40 @@ func (c *Client) addNetworkDevice(
...
+}
+
+// VerifyMAC verifies that the MAC is valid for VMWare.
+func VerifyMAC(mac string) bool {
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm.go]: We are missing a uniqueness check, which is important to check / confirm before assignment.

File: provider/vsphere/internal/vsphereclient/createvm_test.go [hunk 7]
```diff
@@ -384,6 +385,35 @@ func (s *clientSuite) TestCreateVirtualMachineRootDiskSize(c *gc.C) {
 	})
 }
 
+func (s *clientSuite) TestVerifyMAC(c *gc.C) {
+	var testData = []struct {
+		Mac    string
+		Result bool
+	}{
+		{"foo:bar:baz", false},
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm_test.go]: If we're going with a fixed set of cases here, let's validate all the edges. For instance I notice we're not playing with mac that have letters outside of a-f, nor having letters outside the right locations, nor going over the max value, etc.

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 8]
```diff
@@ -469,6 +466,13 @@ func (c *Client) addNetworkDevice(
 	wakeOnLan := true
 	networkDevice.WakeOnLanEnabled = &wakeOnLan
 	networkDevice.Backing = networkBacking
+	if mac != "" {
+		if !VerifyMAC(mac) {
+			return nil, fmt.Errorf("Invalid MAC address: %q", mac)
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm.go]: Let's return a short snippet of what a valid mac looks like. Also consider adding snippets to verifymac and return specific errors on where it failed validation.

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 9]
```diff
@@ -24,6 +28,19 @@ import (
 
 //go:generate go run ../../../../generate/filetoconst/filetoconst.go UbuntuOVF ubuntu.ovf ovf_ubuntu.go 2017 vsphereclient
 
+// NetworkDevice defines a single network device attached to a newly created VM.
+type NetworkDevice struct {
+	// Network is the name of the network the device should be connected to.
+	// If empty it will be connected to the default "VM Network" network.
+	Network string
+	// MAC is the hardware address of the network device. It has to be
+	// in the form of
```
Comment (nskaggs) [inline provider/vsphere/internal/vsphereclient/createvm.go]: in the form of ...?

File: provider/vsphere/environ_broker.go [hunk 10]
```diff
@@ -137,18 +139,41 @@ func (env *sessionEnviron) newRawInstance(
...
-		cloudcfg.AddBootCmd(commands...)
+		interfaces = append(interfaces, network.InterfaceInfo{
+			InterfaceName: "eth1",
+			MACAddress:    externalMac,
+			InterfaceType: network.EthernetInterface,
+			ConfigType:    network.ConfigDHCP,
+		})
+		networkDevices = append(networkDevices, vsphereclient.NetworkDevice{MAC: externalMac, Network: externalNetwork})
+	}
+	// TODO(wpk) There's no (known) way to tell cloud-init to disable network (using cloudinit.CloudInitNetworkConfigDisabled)
```
Comment (nskaggs) [inline provider/vsphere/environ_broker.go]: Do we have an upstream bug for this?

File: provider/vsphere/environ_broker.go [hunk 11]
```diff
@@ -137,18 +139,41 @@ func (env *sessionEnviron) newRawInstance(
...
+	// so the network might be double-configured. That should be ok as long as we're using DHCP.
+	err = cloudcfg.AddNetworkConfig(interfaces)
```
Comment (nskaggs) [inline provider/vsphere/environ_broker.go]: Is this the first place we'd error if the generated mac is not unique? I'm trying to think through how juju might generate invalid config, and make sure we catch it ourselves and deal with it, as opposed to handing off to the user.

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        if replace_ethernets(args.intf_file, ip_output, (tries != retries - 1)):
+             break
+        else:
+             time.sleep(float(args.wait))
+
+if __name__ == "__main__":
+    main()
+`
+
+const CloudInitNetworkConfigDisabled = `network:
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: AFAICT, this should have no network config set by cloud-init right? I saw other comments this wasn't working?

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+
+			dnsSearchDomains := strings.Join(prepared.DNSSearchDomains, " ")
+			if dnsSearchDomains != "" {
+				output.WriteString("  dns-search ")
+				output.WriteString(dnsSearchDomains + "\n")
+			}
+			continue
+		}
+
+		address, hasAddress := prepared.NameToAddress[name]
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: It seems like name == "lo" executes through this needlessly. Can we avoid that?

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+			hasV4Interface = true
+		} else {
+			output.WriteString("iface " + name + " inet6 static\n")
+			hasV6Interface = true
+		}
+		output.WriteString("  address " + address + "\n")
+
+		if isIpv4 {
+			if !gateway4Handled && prepared.Gateway4Address != "" {
+				gatewayIP := net.ParseIP(prepared.Gateway4Address)
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: We should only have to parse this gatewayIP once, not every loop.

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+		if info.GatewayAddress.Value != "" {
+			switch {
+			case info.GatewayAddress.Type == network.IPv4Address:
+				iface.Gateway4 = info.GatewayAddress.Value
+			case info.GatewayAddress.Type == network.IPv6Address:
+				iface.Gateway6 = info.GatewayAddress.Value
+			}
+		}
+
+		if info.MTU != 0 && info.MTU != 1500 {
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: Why are we ignoring max MTU size?

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+	nameToAddress := make(map[string]string)
+	nameToRoutes := make(map[string][]network.Route)
+	nameToMTU := make(map[string]int)
+
+	// Always include the loopback.
+	namesInOrder[0] = "lo"
+	autoStarted := set.NewStrings("lo")
+
+	// We need to check if we have a host-provided default GW and use it.
+	// Otherwise we'll use the first device with a gateway address,
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: What if we have multiple devices with a gateway address? It seems possible we can set the wrong default gateway -- can we do better?

File: cloudconfig/cloudinit/network_ubuntu.go [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+		}
+		cfg.AddBootTextFile(jujuNetplanFile, netPlan, 0644)
+		cfg.AddBootTextFile(systemNetworkInterfacesFile+".templ", eni, 0644)
+		cfg.AddBootTextFile(systemNetworkInterfacesFile+".py", NetworkInterfacesScript, 0744)
+		cfg.AddBootCmd(populateNetworkInterfaces(systemNetworkInterfacesFile))
+	}
+	return nil
+}
+
+// Note: we sleep to mitigate against LP #1337873 and LP #1269921.
```
Comment (nskaggs) [inline cloudconfig/cloudinit/network_ubuntu.go]: LP #1269921 seems still valid (though perhaps not on bionic), but LP #1337873 shouldn't be bothering us anymore.

File: cloudconfig/cloudinit/cloudinit_centos.go (hunk 1 above)
Comment (wupeka) [reply cloudconfig/cloudinit/cloudinit_centos.go]: Yes, we can generate /etc/network/ifcfg-ethX for Centos - but IMHO until someone asks as to do so (or someone is bored has too much time to spare) it's not worth doing.

File: cloudconfig/cloudinit/cloudinit_win.go (hunk 2 above)
Comment (wupeka) [reply cloudconfig/cloudinit/cloudinit_win.go]: That's just code refactoring - I don't want to change the logic in any way that's not needed for this PR.

File: provider/vsphere/internal/vsphereclient/createvm_test.go (hunk 3 above)
Comment (wupeka) [reply provider/vsphere/internal/vsphereclient/createvm_test.go]: I just want to be sure that we won't have a situation in which someone in the future decides that 4M MACs is not enough and we should make more bits random.

File: provider/vsphere/internal/vsphereclient/createvm.go (hunk 6 above)
Comment (wupeka) [reply provider/vsphere/internal/vsphereclient/createvm.go]: We don't have anything to compare it to - having two identical MACs in two networks is not that bad, having two identical MACs in a single network would be a problem.

File: provider/vsphere/internal/vsphereclient/createvm.go (hunk 5 above)
Comment (wupeka) [reply provider/vsphere/internal/vsphereclient/createvm.go]: I don't understand - it's currently used only in GenerateMAC and VerifyMAC, and in two very different forms. How would you like to 'unify' it?

**Comment by wupeka** [comment]
$$merge$$

**Comment by wupeka** [comment]
$$merge$$

**Pull request merged**